deepr/example/layers/transformer_movie_length.py

```diff
@@ -0,0 +1,530 @@
+        self.block_id = block_id
+
+    def forward(self, tensors: Tuple[tf.Tensor, tf.Tensor], mode: str = None):
+        with tf.variable_scope(f"multihead_attention_{self.block_id}"):
+            input_enc, mask = tensors
+            queries = keys = values = input_enc
+            batch_size = tf.shape(input_enc)[0]
+
+            Q = tf.layers.dense(
+                queries, self.d_model, use_bias=False, trainable=True, kernel_initializer=None
+            )  # (bs, sl, d_model)
+
+            K = tf.layers.dense(
+                keys, self.d_model, use_bias=False, trainable=True, kernel_initializer=None
+            )  # (bs, sl, d_model)
+
+            V = tf.layers.dense(
+                values, self.d_model, use_bias=False, trainable=True, kernel_initializer=None
+            )  # (bs, sl, d_model)
+
+            Q_ = self.split_heads(Q, batch_size, self.heads, self.d_model)  # (bs, heads, sl, d_model)
+            K_ = self.split_heads(K, batch_size, self.heads, self.d_model)  # (bs, heads, sl, d_model)
+            V_ = self.split_heads(V, batch_size, self.heads, self.d_model)  # (bs, heads, sl, d_model)
+
+            scaled_attention = self.scaled_dot_product_attention(Q_, K_, V_, mask)  # (bs, heads, sl, d_model/heads)
+
+            # restore shape
+            outputs = tf.transpose(scaled_attention, perm=[0, 2, 1, 3])  # (bs, sl, heads, d_model/heads)
+            outputs = tf.reshape(outputs, (batch_size, -1, self.d_model))  # (bs, sl, d_model)
+
+            # project the last dim to `emb_dim` size
+            outputs = tf.layers.dense(outputs, self.emb_dim)  # (bs, sl, emb_dim)
+
+            # residual connection
+            if self.residual_connection:
+                outputs += queries
+
+            if self.use_layer_normalization:
+                outputs = NormalizationLayer()(outputs, mode)
+
+            return outputs  # (bs, sl, emb_dim)
+
+    @staticmethod
+    def split_heads(x, batch_size, heads, d_model):
+        """ Split the last dimension into heads """
+        x = tf.reshape(x, (batch_size, -1, heads, d_model // heads))
+        return tf.transpose(x, perm=[0, 2, 1, 3])
+
+    @staticmethod
+    def scaled_dot_product_attention(query, key, value, mask=None):
+        """Compute 'Scaled Dot Product Attention'"""
+        with tf.variable_scope("scaled_dot_product_attention"):
+            d_k = tf.cast(tf.shape(key)[-1], tf.float32)
+            scores = tf.matmul(query, key, transpose_b=True, name="DENI_1")
+            scores /= tf.math.sqrt(d_k)  # (bs, heads, sl, sl)
+            if mask is not None:
+                scores += mask * -1e9  # the masked values will be replaced by -1e9
+            attention_weights = tf.nn.softmax(scores, axis=-1, name="attention_weights")  # (bs, heads, sl, sl)
+            return tf.matmul(attention_weights, value, name="DENI_2")  # (bs, heads, sl, d_model/heads)
+
+
+def main():
+    # 1. load dataset
+    train_dataset, validation_dataset = prepare_dataset(
```

Comment:
using https://www.tensorflow.org/datasets/catalog/movie_lens might be more convenient ?

Reply:
yes, it's much more convenient! Thank you, I didn't know about tf build-in datasets